File: tools/packages/codeflow-utils-python/src/codeflow_utils/formatting/number.py
"""Number formatting utilities."""


def format_number(
    value: float | int,
    decimals: int = 2,
    thousands_separator: str = ",",
    decimal_separator: str = ".",
) -> str:
    """
    Format number with thousands separator and decimal places.

    Args:
        value: Number to format
        decimals: Number of decimal places
        thousands_separator: Thousands separator character
        decimal_separator: Decimal separator character

    Returns:
        Formatted number string
    """
    # Format with decimal places
    formatted = f"{value:,.{decimals}f}"

    # Replace separators if needed
    if thousands_separator != ",":
        formatted = formatted.replace(",", "TEMP_THOUSANDS")
        formatted = formatted.replace(".", decimal_separator)
        formatted = formatted.replace("TEMP_THOUSANDS", thousands_separator)
    elif decimal_separator != ".":
        # Split by decimal point
        parts = formatted.split(".")
        if len(parts) == 2:
            formatted = thousands_separator.join(parts[0].split(",")) + decimal_separator + parts[1]
        else:
            formatted = thousands_separator.join(parts[0].split(","))

    return formatted


def format_bytes(bytes_value: int, binary: bool = False) -> str:
    """
    Format bytes to human-readable string.

    Args:
        bytes_value: Number of bytes
        binary: Use binary (1024) or decimal (1000) units

    Returns:
        Formatted string (e.g., "1.5 MB")
    """
    base = 1024 if binary else 1000
    units = ["B", "KB", "MB", "GB", "TB", "PB"] if not binary else ["B", "KiB", "MiB", "GiB", "TiB", "PiB"]

    if bytes_value == 0:
        return "0 B"

    unit_index = 0
    value = float(bytes_value)

    while value >= base and unit_index < len(units) - 1:
        value /= base
        unit_index += 1

    return f"{value:.2f} {units[unit_index]}"


def format_percentage(value: float, decimals: int = 1) -> str:
    """
    Format number as percentage.

    Args:
        value: Number to format (0.0 to 1.0 or 0 to 100)
        decimals: Number of decimal places

    Returns:
        Formatted percentage string
    """
    # Assume value is 0-1 if less than 1, otherwise 0-100
    if value <= 1.0:
        percentage = value * 100
    else:
        percentage = value

    return f"{percentage:.{decimals}f}%"

File: desktop/packages/utils/src/formatting/number.ts
/**
 * Number formatting utilities.
 */

/**
 * Format number with thousands separator and decimal places.
 *
 * @param value - Number to format
 * @param decimals - Number of decimal places
 * @param thousandsSeparator - Thousands separator character
 * @param decimalSeparator - Decimal separator character
 * @returns Formatted number string
 */
export function formatNumber(
  value: number,
  decimals: number = 2,
  thousandsSeparator: string = ",",
  decimalSeparator: string = "."
): string {
  return value.toLocaleString("en-US", {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  })
    .replace(/,/g, thousandsSeparator)
    .replace(/\./g, decimalSeparator);
}

/**
 * Format bytes to human-readable string.
 *
 * @param bytesValue - Number of bytes
 * @param binary - Use binary (1024) or decimal (1000) units
 * @returns Formatted string (e.g., "1.5 MB")
 */
export function formatBytes(bytesValue: number, binary: boolean = false): string {
  const base = binary ? 1024 : 1000;
  const units = binary
    ? ["B", "KiB", "MiB", "GiB", "TiB", "PiB"]
    : ["B", "KB", "MB", "GB", "TB", "PB"];

  if (bytesValue === 0) {
    return "0 B";
  }

  let unitIndex = 0;
  let value = bytesValue;

  while (value >= base && unitIndex < units.length - 1) {
    value /= base;
    unitIndex++;
  }

  return `${value.toFixed(2)} ${units[unitIndex]}`;
}

/**
 * Format number as percentage.
 *
 * @param value - Number to format (0.0 to 1.0 or 0 to 100)
 * @param decimals - Number of decimal places
 * @returns Formatted percentage string
 */
export function formatPercentage(value: number, decimals: number = 1): string {
  const percentage = value <= 1.0 ? value * 100 : value;
  return `${percentage.toFixed(decimals)}%`;
}
